### PR TITLE
various enhancements

### PR DESCRIPTION
**Pull request by cppforlife**
review via commits view

### CODE REVIEW

File: cmd/controller/run.go [hunk 1]
```diff
@@ -164,7 +164,7 @@ func Run(opts Options, runLog logr.Logger) {
 	{ // add controller for PackageInstall
 		pkgInstallCtrlOpts := controller.Options{
 			Reconciler:              NewPackageInstallReconciler(kcClient, pkgClient, coreClient, runLog.WithName("ipr")),
-			MaxConcurrentReconciles: opts.Concurrency,
+			MaxConcurrentReconciles: 1,
```
Comment (danielhelfand) [inline cmd/controller/run.go:167]: What is the motivation behind this change?

Wouldn't this be a breaking change if we remove this? Should we maybe consider a formal deprecation process?

File: pkg/packageinstall/packageinstall.go [hunk 2]
```diff
@@ -28,6 +28,13 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
+const (
+	// DowngradableAnnKey specifies annotation that user can place on
```
Comment (joe-kimmel-vmw) [inline pkg/packageinstall/packageinstall.go:32]: in this diff it's easy to see how this gets used as `canDowngrade` so it's just a boolean flag.  Would we ever allow someone to specify a version floor,  e.g. `"canDowngradeUntil": "1.0.1"`. indicating that they won't accept a version lower than the value?

File: pkg/apis/packaging/v1alpha1/package_install.go [hunk 3]
```diff
@@ -93,4 +93,8 @@ type PackageInstallStatus struct {
 	// TODO this is desired resolved version (not actually deployed)
 	// +optional
 	Version string `json:"version,omitempty"`
+	// LastAttemptedVersion specifies what version was last attempted to be installed.
+	// It does _not_ indicate it was successfully installed.
+	// +optional
+	LastAttemptedVersion string `json:"lastAttemptedVersion,omitempty"`
```
Comment (danielhelfand) [inline pkg/apis/packaging/v1alpha1/package_install.go:99]: Would we want to eventually expand this and make it a struct that would also include if it was successfully installed? Should there be a follow up issue here?

File: pkg/packageinstall/packageinstall.go [hunk 4]
```diff
@@ -284,25 +355,25 @@ func (pi *PackageInstallCR) update(updateFunc func(*pkgingv1alpha1.PackageInstal
 }
 
 func (pi *PackageInstallCR) reconcileFetchPlaceholderSecrets(pkg datapkgingv1alpha1.Package) (datapkgingv1alpha1.Package, error) {
-	pkgCopy := pkg.DeepCopy()
-	for i, fetch := range pkgCopy.Spec.Template.Spec.Fetch {
+	pkg = *pkg.DeepCopy()
```
Comment (joe-kimmel-vmw) [inline pkg/packageinstall/packageinstall.go:358]: imo it's more clear to name it `pkgCopy`  as it eliminates all possible confusions

File: pkg/packageinstall/packageinstall.go [hunk 5]
```diff
@@ -91,9 +98,34 @@ func (pi *PackageInstallCR) reconcile(modelStatus *reconciler.Status) (reconcile
 		return reconcile.Result{Requeue: true}, err
 	}
 
+	// Set new desired version before checking if it's not applicable
 	pi.model.Status.Version = pkg.Spec.Version
 
-	existingApp, err := pi.kcclient.KappctrlV1alpha1().Apps(pi.model.Namespace).Get(context.Background(), pi.model.Name, metav1.GetOptions{})
+	_, canDowngrade := pi.model.Annotations[DowngradableAnnKey]
+	if !canDowngrade && len(pi.model.Status.LastAttemptedVersion) > 0 {
```
Comment (danielhelfand) [inline pkg/packageinstall/packageinstall.go]: Would `pi.model.Status.LastAttemptedVersion != ""` read a bit clearer here?

File: pkg/packageinstall/packageinstall_downgrade_test.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+		assert.Equal(t, "ver-1.0.0", getApp(t, appClient, "instl-pkg").Spec.Fetch[0].ImgpkgBundle.Image)
+	})
+}
+
+func getPackageInstall(t *testing.T, clientset *fakekappctrl.Clientset, name string) *pkgingv1alpha1.PackageInstall {
+	gvr := schema.GroupVersionResource{"packaging.carvel.dev", "v1alpha1", "packageinstalls"}
+	obj, err := clientset.Tracker().Get(gvr, "", name)
+	assert.Nil(t, err)
+	assert.NotNil(t, obj)
```
Comment (danielhelfand) [inline pkg/packageinstall/packageinstall_downgrade_test.go]: I think `require.NotNil` would be helpful here to avoid nil being returned

File: pkg/packageinstall/packageinstall_downgrade_test.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+	return obj.(*pkgingv1alpha1.PackageInstall)
+}
+
+func getApp(t *testing.T, clientset *fakekappctrl.Clientset, name string) *v1alpha1.App {
+	gvr := schema.GroupVersionResource{"kappctrl.k14s.io", "v1alpha1", "apps"}
+	obj, err := clientset.Tracker().Get(gvr, "", name)
+	assert.Nil(t, err)
+	assert.NotNil(t, obj)
```
Comment (danielhelfand) [inline pkg/packageinstall/packageinstall_downgrade_test.go]: I think `require.NotNil` would be helpful here to avoid nil being returned

**Review by joe-kimmel-vmw — COMMENTED** [review]
just one general question and a nit re. naming things

File: cmd/controller/run.go (hunk 1 above)
Comment (cppforlife) [reply cmd/controller/run.go:167]: opts.Concurrency is used for App CRs. probably was copy pasted for PackageInstalls.

> Wouldn't this be a breaking change if we remove this?

im not sure i see what it breaks.

File: pkg/packageinstall/packageinstall.go (hunk 4 above)
Comment (cppforlife) [reply pkg/packageinstall/packageinstall.go:358]: it leads to potential of modifying and/or returning the wrong thing.

File: pkg/apis/packaging/v1alpha1/package_install.go (hunk 3 above)
Comment (cppforlife) [reply pkg/apis/packaging/v1alpha1/package_install.go:99]: i would say not... at least at this point. if we would want to track last successful version, then we would probably introduce status.LastSuccessfullyReconciledVersion.

File: pkg/packageinstall/packageinstall.go (hunk 2 above)
Comment (cppforlife) [reply pkg/packageinstall/packageinstall.go:32]: yeah i was thinking to support that, but i couldnt find concrete use cases. the main goal is to allow spec.versionSelection to be able to select version lower than what was previously attempted. ultimately spec.versionSelection is the thing that specifies version floor of whats allowed.

File: pkg/packageinstall/packageinstall_downgrade_test.go (hunk 6 above)
Comment (cppforlife) [reply pkg/packageinstall/packageinstall_downgrade_test.go]: fixed.

File: pkg/packageinstall/packageinstall_downgrade_test.go (hunk 7 above)
Comment (cppforlife) [reply pkg/packageinstall/packageinstall_downgrade_test.go]: fixed.

File: pkg/packageinstall/packageinstall.go (hunk 5 above)
Comment (cppforlife) [reply pkg/packageinstall/packageinstall.go]: updated.

**Review by joe-kimmel-vmw — APPROVED** [review]

File: cmd/controller/run.go (hunk 1 above)
Comment (danielhelfand) [reply cmd/controller/run.go:167]: Are we treating this as a bug then? My thought is that this is a configurable property that is now being made unconfigurable and that is what is considered breaking.

Comment (danielhelfand) [reply cmd/controller/run.go:167]: > i should have added to the comment: currently it seems that multiple items will get reconciled in parallel. that i think also includes same items which would in theory introduce conflicts in updates. i just set it to 1 explicitly (that's actually controller runtime default afaik) because unlike App reconciler, reconciling PackageInstall should take millis/seconds.

This is considered not breaking in follow up convo. Closing this out.

**Review by danielhelfand — APPROVED** [review]
lgtm
